internal/langserver/handlers/did_open.go

```diff
@@ -2,19 +2,35 @@ package handlers
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/creachadair/jrpc2"
 	"github.com/hashicorp/terraform-ls/internal/document"
 	"github.com/hashicorp/terraform-ls/internal/job"
-	ilsp "github.com/hashicorp/terraform-ls/internal/lsp"
 	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
 	"github.com/hashicorp/terraform-ls/internal/state"
 	"github.com/hashicorp/terraform-ls/internal/terraform/exec"
 	"github.com/hashicorp/terraform-ls/internal/terraform/module"
 	op "github.com/hashicorp/terraform-ls/internal/terraform/module/operation"
+	"github.com/hashicorp/terraform-ls/internal/uri"
 )
 
 func (svc *service) TextDocumentDidOpen(ctx context.Context, params lsp.DidOpenTextDocumentParams) error {
-	dh := ilsp.HandleFromDocumentURI(params.TextDocument.URI)
+	docURI := string(params.TextDocument.URI)
+
+	// URIs are always checked during initialize request, but
+	// we still allow single-file mode, therefore invalid URIs
+	// can still land here, so we check for those.
+	if !uri.IsURIValid(docURI) {
+		jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+			Type: lsp.Warning,
+			Message: fmt.Sprintf("Ignoring workspace folder (unsupport or invalid URI) %s."+
+				" This is most likely bug, please report it.", docURI),
+		})
+		return fmt.Errorf("invalid URI: %s", docURI)
```

Comment:
Won't this open two message windows (one for warning and one for error)?

Reply:
Hmm, not sure actually - good question! It depends on how we deal with error from `textDocument/didOpen`. I'll test this scenario before merging.